src/compute/mod.rs

```diff
@@ -10,121 +10,214 @@ pub(crate) mod grid;
     {
         #[cfg(feature = "debug")]
-        NODE_LOGGER.labelled_debug_log("CACHE", cached_size);
-        #[cfg(feature = "debug")]
-        NODE_LOGGER.labelled_debug_log("run_mode", run_mode);
-        #[cfg(feature = "debug")]
-        NODE_LOGGER.labelled_debug_log("sizing_mode", sizing_mode);
-        #[cfg(feature = "debug")]
-        NODE_LOGGER.labelled_debug_log("known_dimensions", known_dimensions);
+        NODE_LOGGER.labelled_debug_log("CACHE", cached_size_and_baselines.size);
         #[cfg(feature = "debug")]
-        NODE_LOGGER.labelled_debug_log("available_space", available_space);
+        debug_log_node(known_dimensions, parent_size, available_space, run_mode, sizing_mode);
         #[cfg(feature = "debug")]
         NODE_LOGGER.pop_node();
-        return cached_size;
+        return cached_size_and_baselines;
     }
 
     #[cfg(feature = "debug")]
     NODE_LOGGER.log("COMPUTE");
     #[cfg(feature = "debug")]
-    NODE_LOGGER.labelled_debug_log("run_mode", run_mode);
-    #[cfg(feature = "debug")]
-    NODE_LOGGER.labelled_debug_log("sizing_mode", sizing_mode);
-    #[cfg(feature = "debug")]
-    NODE_LOGGER.labelled_debug_log("known_dimensions", known_dimensions);
-    #[cfg(feature = "debug")]
-    NODE_LOGGER.labelled_debug_log("available_space", available_space);
-
-    // If this is a leaf node we can skip a lot of this function in some cases
-    let computed_size = if tree.is_childless(node) {
+    debug_log_node(known_dimensions, parent_size, available_space, run_mode, sizing_mode);
+
+    /// Inlined function generic over the LayoutAlgorithm to reduce code duplication
+    #[inline(always)]
+    fn perform_computations<Algorithm: LayoutAlgorithm>(
+        tree: &mut impl LayoutTree,
+        known_dimensions: Size<Option<f32>>,
+        parent_size: Size<Option<f32>>,
+        available_space: Size<AvailableSpace>,
+        run_mode: RunMode,
+        sizing_mode: SizingMode,
+    ) -> SizeAndBaselines {
         #[cfg(feature = "debug")]
-        NODE_LOGGER.log("Algo: leaf");
-        self::leaf::compute(tree, node, known_dimensions, parent_size, available_space, run_mode, sizing_mode)
-    } else {
-        // println!("match {:?}", tree.style(node).display);
-        match tree.style(node).display {
-            Display::Flex => {
-                #[cfg(feature = "debug")]
-                NODE_LOGGER.log("Algo: flexbox");
-                self::flexbox::compute(tree, node, known_dimensions, parent_size, available_space, run_mode)
+        NODE_LOGGER.log(Algorithm::NAME);
+
+        match run_mode {
+            RunMode::PeformLayout => {
+                Algorithm::perform_layout(tree, known_dimensions, parent_size, available_space, sizing_mode)
             }
+            RunMode::ComputeSize => {
+                let size = Algorithm::measure_size(tree, known_dimensions, parent_size, available_space, sizing_mode);
+                SizeAndBaselines { size, first_baselines: Point::NONE }
```

Comment:
If this only returns the first baseline and not multiple, should it be named "SizeAndBaseline"?

Reply:
Well it (in theory at least) returns a baseline in each dimension. We may also want to extend it in future to return a last baseline in addition to a first baseline. And possibly to return different baselines depending on the script being used (latin, arabic, etc). See https://www.w3.org/TR/css-align-3/#baseline-rules

Reply:
Also note: the baseline stuff is really part of #325. This PR is just based on top of that one.